tests/test_server.py

```diff
@@ -245,6 +246,28 @@ def test_get_all_shopcart(self):
         carts = json.loads(resp.data.decode('utf8'))
         self.assertEqual( len(carts), 3)
 
+    def test_query_product(self):
+        cart = server.Shopcart.find(1)
+        cart.add_product( 5 )
+        cart.add_product( 6 )
+        cart.save() 
+        cart = server.Shopcart.find(2)
+        cart.add_product( 6 )
+        cart.save()
+        self.assertEqual(len(server.Shopcart.find_by_product(123)), 0)
+        # ?productid=123
+        resp = self.app.get('/shopcarts', query_string='pid=123' )
+        self.assertEqual( resp.status_code, status.HTTP_200_OK )
+        data = json.loads(resp.data.decode('utf8'))
+        self.assertEqual(data, [])
+        
+        #count for 5 should have cart 1, 4
+        resp = self.app.get('/shopcarts', query_string='pid=5' )
```

Comment:
I would probably break these out into two separate tests here: 1. For query for products that exist in some carts and 2. query a product `123` that doesn't exist in any cart.